linkerd/app/inbound/src/direct.rs

```diff
@@ -484,3 +485,21 @@ impl<T> InsertParam<tls::ConditionalServerTls, T> for TlsParams {
         (tls, target)
     }
 }
+
+// === impl Discover ===
+
+impl svc::Param<OrigDstAddr> for Discover {
+    fn param(&self) -> OrigDstAddr {
+        if self.header.name.is_none() {
+            // When the transport header targets an alternate port (but does
+            // not identify an alternate target name), we check the new
+            // target's policy (rather than the inbound proxy's address).
+            return OrigDstAddr((self.client.local_addr.ip(), self.header.port).into());
+        }
+
+        // When the transport header provides an alternate target, the
+        // connection is a gateway connection. We check the _gateway
+        // address's_ policy (rather than the target address).
+        self.client.local_addr
+    }
+}
```

Comment:
This is a bit of a red flag to me... it indicates that we're using the `OrigDstAddr` type incorrectly.

* we should have a newtype that clearly indicates what we're using it for.
* then this impl sorts out whether the orig dst is used or something else.

Also I'm having trouble parsing this. 
```suggestion
impl svc::Param<policy::LookupAddr> for Discover {
    fn param(&self) -> policy::LookupAddr {
        // When the transport header provides an alternate target, the
        // connection is a gateway connection. We use the `OrigDstAddr` to lookup policies.
        if self.header.name.is_some() {
            return policy::LookupAddr(*self.client.local_addr);
        }
        
        // Otherwise, use the port override from the header so we. 
        policy::LookupAddr((self.client.local_addr.ip(), self.header.port).into());
    }
}
```

Reply:
Assuming we never want to turn a `LookupAddr` back into an `OrigDstAddr` (which...kind of defeats the purpose of the separate type, since a `LookupAddr` may not have been an actual original destination), this ends up being kind of a large change --- the address passed to `get_policy` ends up being propagated through a lot of different server policy code. I'm happy to change it, but it's gonna be a much bigger diff.

Reply:
OK let me take a quick look at the current state of things.

Reply:
in particular, it means that all the authz checks would be performed on a `LookupAddr`, which..._also_ smells weird to me. Although, I guess, at that point it could just be turned back into an un-newtyped `SocketAddr`...:woman_shrugging:

Reply:
OK we can fix this newtype a followup.

Reply:
well, i went ahead and did it (1252ca721e027836a516f5d11b5e6cef75fd4f2a), but am happy to back it out if you prefer?

Reply:
as an aside, it occurs to me that this logic could have been an `ExtractParam` impl and that the `Discover` target _could_ go away...not sure if that's nicer in this case though :woman_shrugging:

Reply:
eh, let's keep your change for a new branch and [3a2e376](https://github.com/linkerd/linkerd2-proxy/pull/2186/commits/3a2e376cf9bcffdbea17233f76eeb58ba3cfcc8e) + blanket impl?

After some poking it at, i don't know that we really want a `LookupAddr` type to be ~everywhere in the inbound stack. And it's not really correct for the inbound stack to rely on `OrigDstAddr` either, since that's not really accurate. What we want is:

* `OrigDstAddr` is used **only** to indicate the address returned by the socket option lookup.
* `ServerAddr` is used to refer to the address an inbound connection targets.
  * When we know `ServerAddr` is not the proxy's inbound server, we use `Remote<ServerAddr>`
  * If it's useful, we can describe direct connections targeting a `Local<ServerAddr>`
* The `LookupAddr` type is probably still use as a param marker to indicate when the param controls. It should _probably_ be a `LookupAddr(ServerAddr)`.

Reply:
mmkay, that makes sense — i'll make that change in a separate branch.

Reply:
okay, put everything back the way it was, and addressed the other review feedback. considered adding a `GetPolicy::into_service` as well, but figured we could add that if we actually need it. i'll make a separate PR for the address param changes.

Reply:
pulled the `LookupAddr` change out into a separate PR  #2264